### PR TITLE
[#395] include user_type in permissions

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

**Comment by korydraughn** [comment]
Is this ready for review?

**Comment by d-w-moore** [comment]
@korydraughn - yes!

**Comment by trel** [comment]
is `ils-A-analog.py` a kind of test - to make sure it's working as expected?   It should be morphed into a real test?

File: ils-A-analog.py
Comment (korydraughn) [inline ils-A-analog.py:12]: Does this mean the PRC doesn't provide functions like `is_collection(path)` and `is_data_object(path)`?

File: ils-A-analog.py
Comment (d-w-moore) [reply ils-A-analog.py:1]: I think this will be removed from the final PR, as it was only useful for establishing the erroneous behavior.

Comment (d-w-moore) [reply ils-A-analog.py:1]: @trel   `acls` is shorter and can eventually replace `permissions`.  I did not want to change the value of a default parameter in a way that would likely break a lot of existing code.

**Comment by d-w-moore** [comment]
> Can you talk through the reasoning for adding `.acls` rather than just updating a new default parameter for `.permissions`?

@trel `acls` is shorter and can eventually replace `permissions`. I did not want to change the value of a default parameter in a way that would likely break a lot of existing code.

File: ils-A-analog.py
Comment (d-w-moore) [reply ils-A-analog.py:12]: I'm almost certain it does not. Should it provide those functions as utilities?  It would be easy to do.

File: ils-A-analog.py
Comment (d-w-moore) [reply ils-A-analog.py:20]: It's unnecessary.

File: ils-A-analog.py
Comment (korydraughn) [reply ils-A-analog.py:12]: I think it would be good to have those. Then we wouldn't have to rely on exceptions for this kind of thing.

File: ils-A-analog.py
Comment (trel) [reply ils-A-analog.py:1]: okay, please remove this file and make a proper test that exercises the new code.

**Comment by trel** [comment]
Oh, I see - I missed that `.permissions` already had the option, defined a couple years ago to `False` (in https://github.com/irods/python-irodsclient/commit/7f52a8b3c4e3633161c9bede7f1818312ee33da7) ... and `.acls` would be a shorter way of holding the same call to `.permissions(report_raw_acls = True)`.

Can we enumerate what `report_raw_acls = False` means a bit more in this thread (i.e. with this set of permissions, `False` shows this... and `True` shows this...)

Let's consider how flipping that default will play out - timeline, ramifications, etc.

Not really a fan of having both `.acls` *AND* `.permissions` with different behaviors ... seems semantically overloaded and confusing to a new user.   Ideally we can have only one, and it will have the most common/useful/helpful/correct default.   Just may take time to get there because of backwards compatibility.

**Comment by d-w-moore** [comment]
> Not really a fan of having both `.acls` _AND_ `.permissions` with different behaviors ... seems semantically overloaded and confusing to a new user. Ideally we can have only one, and it will have the most common/useful/helpful/correct default. Just may take time to get there because of backwards compatibility.

Yes, but:

Having the synonym `session.acls`  -- ie. being almost the same as `session.permissions` but it exhibits sane(*) behavior"--  is the least dislikeable solution.

(*) I define "sane" behavior as "**get** reports back the same as what **set** told it to do". That is  the opposite of what `session.permissions` does, albeit the unbalanced behavior is [just for collections, not data objects](https://github.com/irods/python-irodsclient/issues/396).  (Still, pretty darn _confusing_  right?  I mean that's pretty terrible, and that's a big part of what I'm trying to correct here)

Bottom line is I don't think we want to introduce breaking behavior in a micro release, and changing a default parameter within session.permissions would do exactly that.  I agree having these nearly-identical ways of doing one thing is not the greatest, but it doesn't break established code.

If we want to return to "only the one way", then just maybe, we, deprecate `permissions` and remove it it in v1.2.0, or v2.0.0 in favor of `acls`.

Once I've seen it  this way, I can't see it any other way.

**Comment by d-w-moore** [comment]
Besides that, in `ils` lingo we call it an ACL, not a permissions, and the option to ils is, after all, `-A`.  Maybe in PRC, we ultimately decide to be consistent with that ... ?

I mention that not just as snark, but to also to support the argument of (eventually) transitioning `permissions` to `acls`.
```
$ ils -h | grep -i perm
$ ils -h | grep -i acl
 -A  ACL (access control list) and inheritance format

```

**Comment by trel** [comment]
Okay, that's all reasonable.    Let's plan on having both `.acls` and `.permissions` for now, and we'll mark `.permissions` as deprecated.   For 2.0.0 we will change the default and/or remove `.permissions` and/or just make it call `.acls` (please make a new issue for that).

And yes, https://github.com/irods/python-irodsclient/issues/396 is just silly/wrong.

Does that seems good?    I think it satisfies both the 'be sane' and the 'be nice to existing deployments' requirements we seek.

**Comment by d-w-moore** [comment]
@trel - that seems good.  Yes, the permanent solution, when we do it, should be the rename of permissions to acl instead of the current shim.  I'll write up the change in the README too....

**Comment by d-w-moore** [comment]
That seems good.  Yes, the permanent solution, when we do it, should be the rename of permissions to acl instead of the current shim.  

I'll document the changes in the README.

**Comment by alanking** [comment]
Overall this seems fine. If you could drop the latest commit (as suggested here? https://github.com/irods/python-irodsclient/pull/411#discussion_r1026943800) and squash others as appropriate, I'll have another look. Unless these are split up this way for a reason.

File: ils-A-analog.py
Comment (d-w-moore) [reply ils-A-analog.py]: The test now included asserts the conditions looked for (inclusion of only the group ACLs, because they're catalog objects; and exclusion of group members for whom catalog ACLs do not exist).

File: ils-A-analog.py
Comment (d-w-moore) [reply ils-A-analog.py]: This script has been deleted from the PR, and #412 has been made for including commonly needed utility functions such as `is_collection,` etc.

File: README.rst
Comment (trel) [inline README.rst:1136]: "It turns out the the" -> "The"

File: README.rst
Comment (trel) [inline README.rst:1141]: plural?

"in <session>.acl.get(...) method" -> "in <session>.acls.get(...) method"

**Comment by d-w-moore** [comment]
@alanking - Done.

File: README.rst
Comment (d-w-moore) [reply README.rst]: Done

File: README.rst
Comment (d-w-moore) [reply README.rst]: Done

File: README.rst
Comment (trel) [inline README.rst:1141]: since "exception" is a no-no word here... please consider the following instead...

"the exception being that in" -> "except that the"

File: README.rst
Comment (trel) [inline README.rst:1144]: "is now officially deprecated" -> "is deprecated"

File: README.rst
Comment (d-w-moore) [reply README.rst]: Fixed it.

File: README.rst
Comment (d-w-moore) [reply README.rst]: Fixed it...

**Comment by trel** [comment]
please resquash - and then i think i'm good with the changes here.

File: irods/session.py
Comment (korydraughn) [reply irods/session.py:64]: Is `report_raw_acls` about expanding group permissions?

Comment (trel) [reply irods/session.py:64]: yes, `.acls` not having the ability to do the silly thing makes perfect sense to me.  aka, not exposing that knob.

Comment (alanking) [reply irods/session.py:64]: I think I am mainly confused as to what a non-raw-ACL even is. Perhaps the documentation or code comments/docstring should have more of an explanation of what `report_raw_acls` is and what `True` and `False` mean?

I am inclined to agree that the non-raw-ACLs should be hidden in the new interface based on what I've seen so far, but I don't understand it well enough to give an informed opinion. Seeing as how this is a new interface, we don't *need* to expose it, right?

Comment (d-w-moore) [reply irods/session.py:64]: report_raw_acls = False exposes old "faulty/weird" code (which in 7f52a8b3c4e3633161c9bede7f1818312ee33da7 I "replaced" with the newer way of doing things, (although obviously I wasn't bold enough and should have done it more drastically) -- and in the present PR I'm arguing to just take out the option to run the old code in `.acls` ... which will be the eventual replacement for `.permissions`

Comment (d-w-moore) [reply irods/session.py:64]: @alanking Correct, we won't need to expose the old, inconsistent behavior, which was to report "too much" ie more than was in the catalog for iRODS collections.  `.acls` should just tell you what's in the catalog, like ils -A does. And that is what `report_raw_acls` value of `True`  means.

**Comment by d-w-moore** [comment]
> Seems good to me. Is there some way in python to mark something as deprecated other than stating so in the documentation, as you have done?

I don't know of one.  But I'm content with whatever others may suggest.

**Comment by korydraughn** [comment]
If that lib doesn't present any known issues, then I'm all for it.

Ultimately it means rerunning the test suite to make sure introduction of that lib doesn't break anything.

I'm guessing you're referring to https://pypi.org/project/Deprecated/. That was one of the first search results from google.

**Comment by trel** [comment]
yes, a pointer to example usage elsewhere / other projects would be good/nice.

**Comment by trel** [comment]
or... https://pypi.org/project/deprecation/

**Comment by d-w-moore** [comment]
> or... https://pypi.org/project/deprecation/

that is the one I found first.  If I can find a way to deprecate easily without throwing warnings from the base class (session.acls), I'll use one of these.  If not, I'll just record an issue to deprecate `.permissions` in the next release.

**Comment by d-w-moore** [comment]
Oh.  There's already an `exceptions.DeprecationWarning`. Recommended [here](https://dev.to/hckjck/python-deprecation-2mof#class-deprecation)

**Comment by korydraughn** [comment]
By all means, investigate the space and report your findings/thoughts.

**Review by alanking — COMMENTED** [review]
Looks good to me. The use of the keyword for warning suppression is pretty clever (in a good way), IMO. Let's wait for another nod and squash

**Review by korydraughn — APPROVED** [review]
Squash it!

**Comment by d-w-moore** [comment]
@korydraughn done!

**Review by korydraughn — APPROVED** [review]
Let's wait for one more approval before adding the pounds.

**Review by alanking — APPROVED** [review]
Let's do it

**Comment by d-w-moore** [comment]
ok, done and ready.

**Pull request merged**
